enhancement(components): Add customizable connection retry options for Pulsar client in Pulsar sink

FRosner opened the pull request:
I want to be able to customize the connection retry options for the Pulsar client. This fixes https://github.com/vectordotdev/vector/issues/21242

<!--
**Your PR title must conform to the conventional commit spec!**

  <type>(<scope>)!: <description>

  * `type` = chore, enhancement, feat, fix, docs
  * `!` = OPTIONAL: signals a breaking change
  * `scope` = Optional when `type` is "chore" or "docs", available scopes https://github.com/vectordotdev/vector/blob/master/.github/semantic.yml#L20
  * `description` = short description of the change

Examples:

  * enhancement(file source): Add `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fix a bug discovering new files
  * chore(external docs): Clarify `batch_size` option
-->

jszwedko commented on src/sinks/pulsar/config.rs:
```diff
@@ -171,6 +176,47 @@ pub enum PulsarCompression {
+    /// Minimum delay between connection retries.
+    #[configurable(metadata(docs::type_unit = "milliseconds"))]
+    #[configurable(metadata(docs::examples = 10))]
+    pub min_backoff: Duration,
```
```suggestion
    pub min_backoff_ms: Duration,
```

We generally suffix duration types with their unit per https://github.com/vectordotdev/vector/blob/master/docs/specs/configuration.md#option-naming

FRosner replied:
https://github.com/vectordotdev/vector/pull/21245/commits/18c535dfe89382f547d0a700bf38e4eb350f1c39

jszwedko commented on src/sinks/pulsar/config.rs:
```diff
@@ -171,6 +176,47 @@ pub enum PulsarCompression {
+    /// Maximum delay between reconnection retries.
+    #[configurable(metadata(docs::type_unit = "seconds"))]
+    #[configurable(metadata(docs::examples = 30))]
+    pub max_backoff: Duration,
```
```suggestion
    pub max_backoff_secs: Duration,
```

FRosner replied:
https://github.com/vectordotdev/vector/pull/21245/commits/18c535dfe89382f547d0a700bf38e4eb350f1c39

jszwedko commented on src/sinks/pulsar/config.rs:
```diff
@@ -171,6 +176,47 @@ pub enum PulsarCompression {
+    /// Time limit to establish a connection.
+    #[configurable(metadata(docs::type_unit = "seconds"))]
+    #[configurable(metadata(docs::examples = 10))]
+    pub connection_timeout: Duration,
```
```suggestion
    pub connection_timeout_secs: Duration,
```

FRosner replied:
https://github.com/vectordotdev/vector/pull/21245/commits/18c535dfe89382f547d0a700bf38e4eb350f1c39

jszwedko commented on src/sinks/pulsar/config.rs:
```diff
@@ -171,6 +176,47 @@ pub enum PulsarCompression {
+    /// Keep-alive interval for each broker connection.
+    #[configurable(metadata(docs::type_unit = "seconds"))]
+    #[configurable(metadata(docs::examples = 60))]
+    pub keep_alive: Duration,
```
```suggestion
    pub keep_alive_secs: Duration,
```

FRosner replied:
https://github.com/vectordotdev/vector/pull/21245/commits/18c535dfe89382f547d0a700bf38e4eb350f1c39

jszwedko reviewed (commented):
Apologies for the delay in review @FRosner . I think this looks like a pretty straight-forward addition to me. In addition to my few inline comments, could you:

- Add a changelog entry per https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md ?
- Run `make generate-component-docs` to regenerate the component documentation?

Otherwise I think this looks good.

FRosner commented:
Thanks @jszwedko! I addressed your feedback. While doing so, I realized that I should probably make those new options optional. I converted both the entire option object to be optional, but also each field. I removed the custom default and am picking the pulsar defaults where no user provided value exists.

I'm completely new to rust, so please let me know if my changes aren't making any sense :)

aliciascott reviewed (approved):
good for docs

jszwedko reviewed (approved):
Thanks @FRosner ! Agreed, these should be optional.

FRosner commented:
@jszwedko I had to run `cargo fmt` to address the formatting issues. I pushed the changes.
